codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -925,5 +915,295 @@ final void incrementStreamableBytesForTree(int numBytes) {
+         */
+        public final boolean isWritable(AbstractState state) {
+            return isWritableConnection() && state.windowSize() - state.pendingBytes() > 0;
+        }
+
+        protected final void writePendingBytes(Http2StreamVisitor writeAllocatedBytesVisitor) throws Http2Exception {
+            if (ctx == null) {
+                return;
+            }
+            int connectionWindowSize;
+            do {
+                connectionWindowSize = writableBytes(connectionState.windowSize());
+
+                if (connectionWindowSize > 0) {
+                    // Allocate the bytes for the connection window to the streams, but do not write.
+                    allocateBytesForTree(connectionState.stream(), connectionWindowSize);
+                }
+
+                // Now write all of the allocated bytes, must write as there may be empty frames with
+                // EOS = true
+                connection.forEachActiveStream(writeAllocatedBytesVisitor);
+
+                // We must continue writing until the channel is no longer writable. Even though we use an estimate of
+                // how many bytes we can write until the channel becomes unwritable there may be other ChannelHandlers
+                // in the pipeline that reduce the amount of bytes actually written to the channel.
+            } while (connectionState.streamableBytesForTree() > 0 &&
+                     connectionWindowSize > 0 &&
+                     ctx.channel().isWritable());
+        }
+
+        protected final boolean initialWindowSize(int newWindowSize, Http2StreamVisitor writeAllocatedBytesVisitor)
+                throws Http2Exception {
+            if (newWindowSize < 0) {
+                throw new IllegalArgumentException("Invalid initial window size: " + newWindowSize);
+            }
+
+            final int delta = newWindowSize - initialWindowSize;
+            initialWindowSize = newWindowSize;
+            connection.forEachActiveStream(new Http2StreamVisitor() {
+                @Override
+                public boolean visit(Http2Stream stream) throws Http2Exception {
+                    state(stream).incrementStreamWindow(delta);
+                    return true;
+                }
+            });
+
+            if (delta > 0) {
+                // The window size increased, send any pending frames for all streams.
+                writePendingBytes(writeAllocatedBytesVisitor);
+                return false;
+            }
+            return true;
+        }
+
+        protected final boolean isWritableConnection() {
+            return connectionState.windowSize() - totalPendingBytes > 0 && isChannelWritable();
+        }
+    }
+
+    /**
+     * Provides no notification or tracking of writablity changes.
+     */
+    private final class DefaultWritabilityMonitor extends WritabilityMonitor {
+        protected final Http2StreamVisitor writeAllocatedBytesVisitor = new Http2StreamVisitor() {
```

Comment:
Private +1.

> Should writePendingBytes just be implemented as final in the base class?

It could be if we added a `protected abstract Http2StreamVisitor writeAllocatedBytesVisitor()` (which I had at one point) but I'm not sure if there is any real benefit to this as the base classes would have to override these methods anyways.